File: VirtualWorld-Python/utils/FileIO.py
from zwierzeta.Czlowiek import Czlowiek
from zwierzeta.Antylopa import Antylopa
from zwierzeta.Cyberowca import Cyberowca
from zwierzeta.Lis import Lis
from zwierzeta.Owca import Owca
from zwierzeta.Wilk import Wilk
from zwierzeta.Zolw import Zolw

from rosliny.Barszcz import Barszcz
from rosliny.Guarana import Guarana
from rosliny.Jagody import Jagody
from rosliny.Mlecz import Mlecz
from rosliny.Trawa import Trawa


class FileIO():
    

    def makeSave(self, x, y, organizmy, turaUmiejetnosc, czyCzlowiekIstnieje, czyJestTarczaAldura, kierunekRuchu, obecnaTura, log, prevTura):

        file = open(r"save/save.txt", "w")

        czyCzlowiekIstniejeInt = 1 if czyCzlowiekIstnieje[0] else 0
        czyJestTarczaAldura = 1 if czyJestTarczaAldura else 0

        file.write("x " + str(x) + '\n')
        file.write("y " + str(y) + '\n')
        file.write("t " + str(obecnaTura) + '\n')
        file.write("c " + str(turaUmiejetnosc) + " " + str(kierunekRuchu) + " " + str(czyJestTarczaAldura) + " " + str(czyCzlowiekIstniejeInt) + '\n')

        for i in range(0, len(organizmy)):

            for j in range(0, len(organizmy[i])):

                temp = organizmy[i][j]

                file.write("o " + temp.getZnak() + " " + str(temp.getPolozenieI(0)) + " " + str(temp.getPolozenieI(1)) + " " + str(temp.getSila()) + " " + str(temp.getInicjatywa()) + " " + str(temp.getWiek()) + '\n')

        for i in range(0, len(prevTura)):

            for j in range(0, len(prevTura[i])):

                file.write("p " + prevTura[i][j] + '\n')

        for i in range(0, len(log)):

            file.write(log[i])

        file.close()

    #--------------------------------------------

    def readSave(self, swiat, organizmy):

        CREDITS_HEIGHT = 3

        file = open(r"save/save.txt", "r")

        strs = file.readlines()

        swiat.clearLog()

        swiat.clearPrevTura()

        newStrs = []

        for line in strs:

            newStrs.append(line.strip())


        ret = []

        x = 0
        y = 0

        gotX = False
        gotY = False
        alreadySet = False

        planszNew = []
        creditsNew = []
        organizmyNew = []

        nLines = []

        nLine = 0


        for line in newStrs:

            if len(line) <= 0: continue

            nLine += 1

            if gotX and gotY and alreadySet == False:

                for i in range(0, y):
                    organizmyNew.append([])

                for i in range(0, y):
                    str = "".join(' ' for i in range(x))
                    str = list(str)

                    planszNew.append(str)

                for i in range(0, CREDITS_HEIGHT):

                    if not (i % 2 > 0):

                        str = "".join('-' for i in range(x + 1))
                        str = str + '\n'
        
                        creditsNew.append(str)

                    else:

                        temp = "s188708 Michal Mroz \n"
                        creditsNew.append(temp)



                for i in range(0, swiat.getRozmiarY()):

                    swiat.getPlansza().append([])

                    for j in range(0, swiat.getRozmiarX()):

                        swiat.getPlansza()[i].append(" ")


                alreadySet = True

            
            ch = line[0]

            if ch == 'x':

                x = int(line[2:])
                ret.append(x)
                swiat.changeRozmiarX(x)
                gotX = True

            elif ch == 'y':

                y = int(line[2:])
                ret.append(y)
                swiat.changeRozmiarY(y)
                gotY = True

            elif ch == 't':

                t = int(line[2:])
                swiat.changeObecnaTura(t)

            elif ch == 'c':

                fromIndex = 0

                while line.find(' ', line.find(' ', fromIndex) + 1) != -1:

                    v = int(line[line.find(' ', fromIndex) + 1 : line.find(' ', line.find(' ', fromIndex) + 1)])

                    ret.append(v)

                    fromIndex = line.find(' ', fromIndex) + 1
                
                ret.append(int(line[line.find(' ', fromIndex) + 1 : ]))

            elif ch == 'o':
        
                fi = 2

                values = []

                while line.find(' ', line.find(' ', fi) + 1) != -1:

                    v = int(line[line.find(' ', fi) + 1 : line.find(' ', line.find(' ', fi) + 1)])

                    values.append(v)
                    
                    fi = line.find(' ', fi) + 1

                values.append(int(line[line.find(' ', fi) + 1 : ]))

                tempL = [values[0], values[1]]

                ch2 = line[2]

                temp = Antylopa(swiat, tempL)

                if ch2 == 'A':

                    temp = Antylopa(swiat, tempL)

                elif ch2 == 'B':

                    temp = Barszcz(swiat, tempL)

                elif ch2 == 'C':

                    temp = Czlowiek(swiat, tempL)

                elif ch2 == 'G':

                    temp = Guarana(swiat, tempL)

                elif ch2 == 'J':

                    temp = Jagody(swiat, tempL)

                elif ch2 == 'L':

                    temp = Lis(swiat, tempL)

                elif ch2 == 'M':

                    temp = Mlecz(swiat, tempL)

                elif ch2 == 'O':

                    temp = Owca(swiat, tempL)
                    
                elif ch2 == 'T':

                    temp = Trawa(swiat, tempL)

                elif ch2 == 'W':

                    temp = Wilk(swiat, tempL)

                elif ch2 == 'Z':

                    temp = Zolw(swiat, tempL)

                elif ch2 == 'o':

                    temp = Cyberowca(swiat, tempL)

                temp.changeWiek(values[4])
                temp.changeInicjatywa(values[3])
                temp.zmienSile(values[2])

                planszNew[values[1]][values[0]] = ch2
                organizmyNew[values[1]].append(temp)

            else:
                nLines.append(nLine)


        swiat.changePlansza(planszNew)

        swiat.changeCredits(creditsNew)

        organizmy.clear()

        for i in range(0, len(organizmyNew)):

            organizmy.append([])

        for i in range(0, len(organizmyNew)):

            for j in range(0, len(organizmyNew[i])):

                organizmy[i].append(organizmyNew[i][j])

        i = 0

        j = 0

        nLine = 0

        for n in range(0, len(organizmyNew)):

            swiat.getPrevTura().append([])

        for line in newStrs:

            if len(line) <= 0: continue

            nLine += 1

            if line[0] == 'p':

                temp = line[2:]
                swiat.getPrevTura()[i].append(temp)

                if j + 1 == len(organizmyNew[i]):

                    j = 0
                    i += 1

                else: 

                    j += 1

            elif nLine in nLines:

                temp = line
                if (temp == "Kolejność wykonania ruchów (wiek podczas wykonywania ruchu):"): swiat.addToLog("\n")
                swiat.addToLog(temp)
                

        file.close()

        return ret

        

File: VirtualWorld-Java/src/FileIO.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.concurrent.TimeUnit;

public class FileIO{
	
	public void makeSave(int x, int y, ArrayList<ArrayList<Organizm>> organizmy, int turaUmiejetnosc, Boolean czyCzlowiekIstnieje, Boolean czyJestTarczaAldura, int kierunekRuchu, int obecnaTura) {
		
		try {
			BufferedWriter writer = new BufferedWriter(new FileWriter("save.txt"));
			
			int czyCzlowiekIstniejeInt = czyCzlowiekIstnieje ? 1 : 0;
			int czyJestTarczaAlduraInt = czyJestTarczaAldura ? 1 : 0;
			
			writer.write("x " + Integer.toString(x));
			writer.write("\ny " + Integer.toString(y));
			writer.write("\nt " + Integer.toString(obecnaTura));
			writer.write("\nc " + Integer.toString(turaUmiejetnosc) + " " + Integer.toString(kierunekRuchu) + " " + Integer.toString(czyJestTarczaAlduraInt) + " " + Integer.toString(czyCzlowiekIstniejeInt));
			
			for(int Y=0; Y < organizmy.size(); Y++) {
				for(int X=0; X < organizmy.get(Y).size(); X++) {
					Organizm temp = organizmy.get(Y).get(X);
					
					writer.write("\no " +  temp.getZnak() + " " + temp.getPolozenie(0) + " " + temp.getPolozenie(1) + " " + temp.getSila() + " " + temp.getInicjatywa() + " " + temp.getWiek());
				}
			}
			
			writer.close();
		} 
		
		
		catch(IOException e) {
			e.printStackTrace();
		}
		
	}
	
	public MyResult readSave(Swiat swiat, ArrayList<ArrayList<Organizm>> organizmy) throws InterruptedException {
		File f = new File("save.txt");
			
		if(f.exists() && !f.isDirectory()) {
			
			try {
				BufferedReader reader = new BufferedReader(new FileReader("save.txt"));
				
				String line;
				
				ArrayList<Integer> ret = new ArrayList<Integer>();
				
				int x = 0;
				int y = 0;
				
				Boolean gotX = false;
				Boolean gotY = false;
				Boolean alreadySet = false;
				
				ArrayList<StringBuffer> planszNew = new ArrayList<StringBuffer>();
				
				ArrayList<StringBuffer> creditsNew = new ArrayList<StringBuffer>();
				
				ArrayList<ArrayList<Organizm>> organizmyNew = new ArrayList<ArrayList<Organizm>>();
					
				while((line = reader.readLine()) != null) {
					
					if(gotX && gotY && (alreadySet == false)) {
						
						for(int i=0; i < y; i++) {
							ArrayList<Organizm> temp = new ArrayList<Organizm>();
							organizmyNew.add(temp);
						}
						
						for(int i=0; i<3; i++) {
							if(!(i % 2 > 0)) {
								char[] chars = new char[x + 1];
								Arrays.fill(chars, '-');
								chars[x] = '\n';
								String s = new String(chars);
								StringBuffer temp = new StringBuffer(s);
								creditsNew.add(temp);
							}
							
							else {
								StringBuffer temp = new StringBuffer("s188708 Michal Mroz\n");
								creditsNew.add(temp);
							}
							
						}
						
						
						for(int i=0; i<y; i++) {
							
							StringBuffer temp = new StringBuffer();
							
							for(int j=0; j<x; j++) {
								temp.append(' ');
							}
							
							planszNew.add(temp);
						}
						
						for(int i=0; i<swiat.getRozmiarY(); i++) {
							swiat.getPlansza().add(new StringBuffer());
							for(int j=0; j<swiat.getRozmiarX(); j++) {
								swiat.getPlansza().get(i).append(" ");
							}
						}
											
						alreadySet = true;
					}
					

					switch(line.charAt(0)) {
						case 'x':
							x = Integer.parseInt(line.substring(2));
							ret.add(x);
							swiat.changeRozmiarX(x);
							gotX = true;
							break;
						case 'y':
							y = Integer.parseInt(line.substring(2));
							ret.add(y);
							swiat.changeRozmiarY(y);
							gotY = true;
							break;
						case 't':
							int t = Integer.parseInt(line.substring(2));
							swiat.changeObecnaTura(t);
							break;
						case 'c':
							int fromIndex = 0;
							
							while(line.indexOf(' ', line.indexOf(' ', fromIndex) + 1) != -1) {
								int v = Integer.parseInt(line.substring(line.indexOf(' ', fromIndex) + 1, line.indexOf(' ', line.indexOf(' ', fromIndex) + 1)));
								
								ret.add(v);
								
								fromIndex = line.indexOf(' ', fromIndex) + 1; 
							}
							
							ret.add(Integer.parseInt(line.substring(line.indexOf(' ', fromIndex) + 1)));
							
							break;
							
							
						case 'o':
	
							Organizm temp = null;
							
							int FI = 2;
							
							ArrayList<Integer> values = new ArrayList<Integer>();
							
							while(line.indexOf(' ', line.indexOf(' ', FI) + 1) != -1) {
								int v = Integer.parseInt(line.substring(line.indexOf(' ', FI) + 1, line.indexOf(' ', line.indexOf(' ', FI) + 1)));
								values.add(v);
								FI = line.indexOf(' ', FI) + 1; 
							}
							
							values.add(Integer.parseInt(line.substring(line.indexOf(' ', FI) + 1)));
							
							ArrayList<Integer> tempL = new ArrayList<Integer>();
							tempL.add(values.get(0)); tempL.add(values.get(1));
							
							switch(line.charAt(2)) {
								case 'A':
									temp = new Antylopa(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'B':
									temp = new Barszcz(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'C':
									temp = new Czlowiek(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'G':
									temp = new Guarana(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'J':
									temp = new Jagody(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'L':
									temp = new Lis(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'M':
									temp = new Mlecz(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'O':
									temp = new Owca(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'T':
									temp = new Trawa(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'W':
									temp = new Wilk(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
								case 'Z':
									temp = new Zolw(swiat, tempL, values.get(4), values.get(3), values.get(2));
									break;
							}
							
							planszNew.get(values.get(1)).replace(values.get(0), values.get(0) + 1, ""+line.charAt(2));
							
							organizmyNew.get(values.get(1)).add(temp);
							
							//organizmyNew.get(values.get(1)).get(organizmy.get(values.get(1)).size() - 1).changePolozenie(values.get(0), values.get(1));
							
							break;
					}
					
				}
							
				swiat.changePlansza(planszNew);
				
				swiat.changeCredits(creditsNew);
				
				TimeUnit.SECONDS.sleep(1);
					
				reader.close();
				
				return new MyResult(ret, organizmyNew);
			} 
			
			
			
			catch (IOException e) {
				e.printStackTrace();
				return null;
			}
			
		}
		
		else {
			System.out.println("Zapis nie istnieje!");
			TimeUnit.SECONDS.sleep(1);
			return null;
		}
		
		
	}

}
